From 5070f3343f45925bf3d28bcf4de9c40b52c35e17 Mon Sep 17 00:00:00 2001
From: Louis Marelle <louis.marelle@latmos.ipsl.fr>
Date: Wed, 27 Jul 2022 15:39:34 +0200
Subject: [PATCH 41/52] Force water vapor to be positive in KPP

Modify the WRF-Chem KPP coupler to constrain H20 mixing ratios above 0. Very
weak negative concentrations are sometimes possible when advection is not
positive-definite, and can completely break the chemistry solver.
---
 chem/KPP/util/wkc/gen_kpp_interf_utils.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/chem/KPP/util/wkc/gen_kpp_interf_utils.c b/chem/KPP/util/wkc/gen_kpp_interf_utils.c
index 46161343..8a4ecae2 100644
--- a/chem/KPP/util/wkc/gen_kpp_interf_utils.c
+++ b/chem/KPP/util/wkc/gen_kpp_interf_utils.c
@@ -328,7 +328,7 @@ wki_one_d_vars( FILE * ofile,   knode_t * pp )
   }
 
    fprintf(ofile,"      ! water concentration (molec/cm^3)\n");
-   fprintf(ofile,"    FIX(indf_H2O) = REAL(dens2con_w * moist(i,k,j,P_QV) * rho_phy(i,k,j), KIND=dp)\n\n\n");
+   fprintf(ofile,"    FIX(indf_H2O) = REAL(MAX(dens2con_w * moist(i,k,j,P_QV) * rho_phy(i,k,j), 0.), KIND=dp)\n\n\n");
 
 
 
-- 
2.31.1

